File: js/sort/quickSort.js
// Base Case O(n log n)
// Worst Case O(n ^n)

function quickSort(arr) {
  if (arr.length <= 1) return arr;

  const pivot = arr[arr.length - 1];

  let leftArray = [];
  let righArray = [];

  for (let i = 0; i < arr.length - 1; i++) {
    if (arr[i] >= pivot) {
      righArray.push(arr[i]);
    } else {
      leftArray.push(arr[i]);
    }
  }

  return [...quickSort(leftArray), pivot, ...quickSort(righArray)];
}

console.log(quickSort([-6, 20, 4, 3, 3, 2]));

File: python/sort/quickSort.py
import random


def quickSort(arr: list[int]) -> list[int]:
    if len(arr) <= 1:
        return arr

    pivot = random.choice(arr)
    left = []
    middle = []
    right = []

    for number in arr:
        if number < pivot:
            left.append(number)
        elif number == pivot:
            middle.append(number)
        else:
            right.append(number)

    return quickSort(left) + middle + quickSort(right)


print(quickSort([10, 2, 5, 3, 1, 8, 2]))
